### PR TITLE
Bugfix/agilis 19754 do not keep discovery when timeout passes with dhcp server disabled

### PR DESCRIPTION
**Pull request by metelik**
Finally handling of TIMEOUT and FAIL dhclient -4 's events provided.

### CODE REVIEW

File: lib/nerves_network/dhclientv4.ex [hunk 1]
```diff
@@ -260,10 +274,11 @@ defmodule Nerves.Network.Dhclientv4 do
     {:noreply, state}
   end
 
+  # TIMEOUT event is usualy followed by the (lease)FAIL
```
Comment (adrigonzo) [inline lib/nerves_network/dhclientv4.ex]: typo on "usually"

File: lib/nerves_network/dhcp_manager.ex [hunk 2]
```diff
@@ -44,6 +44,8 @@ defmodule Nerves.Network.DHCPManager do
     # was added to get things going.
     current_interfaces = Nerves.NetworkInterface.interfaces()
 
+    IO.puts "+++++++++++++++ current_interfaces = #{inspect current_interfaces}"
```
Comment (PeterJeaiem) [inline lib/nerves_network/dhcp_manager.ex]: might wanna remove this

Comment (adrigonzo) [inline lib/nerves_network/dhcp_manager.ex]: might need to remove this debug statement.

**Review by PeterJeaiem — APPROVED** [review]

File: lib/nerves_network/dhcp_manager.ex [hunk 3]
```diff
@@ -179,6 +181,46 @@ defmodule Nerves.Network.DHCPManager do
...
+    System.cmd("ping", ["-c", "1", "-q", ipv4_gateway])
+    |> deconfigure_if_gateway_not_pingable(state, info)
+  end
+
+  # This event will be followed by the :leasefail, hence we want the state machine to remain in :dhcp state.
+  # The subsequent :lease fail event shall move the state-machine to the :up state.
+  defp consume(cur_state, {:timeout, info}, state) when cur_state in [:dhcp, :up] do
+    Logger.debug("(context = :dhcp) #{inspect(:timeout)} info: #{inspect(info)}")
+
+    # Let's try newest valid lease. It is an attempt to asnwer whether we are still connected to the same network as before.
```
Comment (adrigonzo) [inline lib/nerves_network/dhcp_manager.ex]: typo on "answer"

File: lib/nerves_network/dhcp_manager.ex [hunk 4]
```diff
@@ -269,21 +311,18 @@ defmodule Nerves.Network.DHCPManager do
...
   ## Neither of these should be called. Not related to Nerves.Network.Dhclientv4.
   defp consume(:dhcp, {:deconfig, _info}, state), do: state
 
-  defp consume(:dhcp, {:leasefail, _info}, state) do
-    dhcp_retry_timer = Process.send_after(self(), :dhcp_retry, state.dhcp_retry_interval)
-
-    %{state | dhcp_retry_timer: dhcp_retry_timer}
-    |> stop_dhclient
-    |> start_link_local
+  defp consume(:dhcp, {reason = :leasefail, _info}, state) do
```
Comment (adrigonzo) [inline lib/nerves_network/dhcp_manager.ex]: reason here will  be an unused variable, you can prob just remove it if unused.

File: lib/nerves_network/dhcp_manager.ex (hunk 3 above)
Comment (metelik) [reply lib/nerves_network/dhcp_manager.ex]: corrected

File: lib/nerves_network/dhcp_manager.ex (hunk 4 above)
Comment (metelik) [reply lib/nerves_network/dhcp_manager.ex]: resolved

File: lib/nerves_network/utils.ex [hunk 5]
```diff
@@ -26,6 +28,47 @@ defmodule Nerves.Network.Utils do
     "169.254.#{x}.#{y}"
   end
 
+  defp bits_set(128), do: 1
```
Comment (adrigonzo) [inline lib/nerves_network/utils.ex]: you can probably make this concise with using a matching clause, like:
`defp bits_set(bits) when bits in [1, 2, 4, 8, 16, 32, 64, 128], do: 1`

File: lib/nerves_network/dhcp_manager.ex (hunk 2 above)
Comment (metelik) [reply lib/nerves_network/dhcp_manager.ex]: removed

Comment (metelik) [reply lib/nerves_network/dhcp_manager.ex]: removed

File: lib/nerves_network/utils.ex (hunk 5 above)
Comment (metelik) [reply lib/nerves_network/utils.ex]: done

**Review by david-naughton — APPROVED** [review]

**Review by adrigonzo — APPROVED** [review]

**Pull request merged**